### PR TITLE
[postgresql] - schema change rework

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #14331 and porting of #17351.

### Summary of Changes

support these schema changes:

- `ALTER TABLE "#__table" ALTER COLUMN "field" SET NOT NULL;`
- `ALTER TABLE "#__table" ALTER COLUMN "field" SET DEFAULT value;`
- `ALTER TABLE "#__table" ALTER COLUMN "field" DROP DEFAULT;`
- `ALTER TABLE "#__table" ALTER COLUMN "field" DROP NOT NULL;`

### Testing Instructions

With latest staging on postgresql go to Manage > Database.

### Expected result
no notice

### Actual result
Notice: Undefined offset:....

@waader , @csthomas  please re-test, review

### CODE REVIEW

**Comment by csthomas** [comment]
1. I have not spent enough time on it but when I changed database table by:
```sql
ALTER TABLE "#__extensions" ALTER COLUMN "custom_data" DROP NOT NULL;
ALTER TABLE "#__extensions" ALTER COLUMN "custom_data" SET DEFAULT '';
```

then joomla does not see any problems, but should show that `DEFAULT` values was removed in
https://github.com/joomla/joomla-cms/blob/0ac5c79bb99f85bb2b62df779d0a6233f664109a/administrator/components/com_admin/sql/updates/postgresql/3.7.0-2017-03-03.sql#L1

2. There is a white space on line 117.

**Comment by alikon** [comment]
1 and 2 fixed

**Comment by alikon** [comment]
reworked a little bit for use `information_schema.columns`

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 1]
```diff
@@ -85,72 +85,85 @@ protected function buildCheckQuery()
...
+						$string = str_replace(';', '', $wordArray[8]);
+						$result = 'SELECT column_name, data_type, is_nullable FROM information_schema.columns WHERE table_name='
+							. $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+							. ' AND (CASE (position(' . $this->db->quote('::') . ' in column_default))'
+							. ' WHEN 0 THEN '
+							. '	column_default = ' . $this->db->quote($string)
+							. ' ELSE '
+							. ' substring(column_default, 1, (position('. $this->db->quote('::') . ' in column_default) -1))  = ' . $this->db->quote($string)
+							. ' END)';
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Lines 139-141 Remove tab

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 2]
```diff
@@ -85,72 +85,85 @@ protected function buildCheckQuery()
...
-					$this->queryType = 'CHANGE_COLUMN_TYPE';
-					$this->checkQueryExpected = 1;
-					$this->msgElements = array($this->fixQuote($wordArray[2]), $this->fixQuote($wordArray[5]), $isNullDef);
+					if ($wordArray[7] . ' ' . $wordArray[8] === 'NOT NULL;')
+					{
+						$isNullable = $this->fixQuote('NO');
+						$result = 'SELECT column_name, data_type, is_nullable , column_default FROM information_schema.columns'
+							. ' WHERE table_name=' . $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+							. ' AND is_nullable = ' . $isNullable;
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Lines 163-165 Remove tab

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 3]
```diff
@@ -85,72 +85,85 @@ protected function buildCheckQuery()
...
-				elseif (strtoupper($wordArray[7] . ' ' . $wordArray[8]) === 'NOT NULL')
-				{
-					if (strtoupper($wordArray[6]) === 'SET')
-					{
-						// SET NOT NULL
-						$isNullable = $this->fixQuote('NO');
-					}
-					else
+					break;
+				case 'SET' :
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Add blank line before

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 4]
```diff
@@ -85,72 +85,85 @@ protected function buildCheckQuery()
...
+						}
 
-					$result = 'SELECT column_name, data_type FROM information_schema.columns WHERE table_name='
-						. $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
-						. ' AND data_type=' . $this->fixQuote($type);
+						$result = 'SELECT column_name, data_type FROM information_schema.columns WHERE table_name='
+							. $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+							. ' AND data_type=' . $this->fixQuote($type);
 
 					$this->queryType = 'CHANGE_COLUMN_TYPE';
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Lines 112-166 indent

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 5]
```diff
@@ -85,72 +85,85 @@ protected function buildCheckQuery()
...
+							. ' ELSE '
+							. ' substring(column_default, 1, (position('. $this->db->quote('::') . ' in column_default) -1))  = ' . $this->db->quote($string)
+							. ' END)';
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
+							$this->checkQueryExpected = 1;
+							$this->msgElements = array($this->fixQuote($wordArray[2]), $this->fixQuote($wordArray[5]), $this->fixQuote($wordArray[8]));
 					}
-					else
+					break;
+
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Remove tabs

**Comment by alikon** [comment]
@Quy hope cs better now

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 6]
```diff
@@ -85,72 +85,86 @@ protected function buildCheckQuery()
...
-				}
-				elseif (strtoupper($wordArray[7] . ' ' . $wordArray[8]) === 'NOT NULL')
-				{
-					if (strtoupper($wordArray[6]) === 'SET')
-					{
-						// SET NOT NULL
+						$this->queryType = 'CHANGE_COLUMN_TYPE';
+						$this->msgElements = array($this->fixQuote($wordArray[2]), $this->fixQuote($wordArray[5]), $type);
+						break;
+					case 'SET' :
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Add a blank line before line 115.

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 7]
```diff
@@ -85,72 +85,86 @@ protected function buildCheckQuery()
...
+						{
+							$isNullable = $this->fixQuote('NO');
+							$result = 'SELECT column_name, data_type, is_nullable , column_default FROM information_schema.columns'
+								. ' WHERE table_name=' . $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+								. ' AND is_nullable = ' . $isNullable;
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
+							$this->checkQueryExpected = 0;
+							$this->msgElements = array($this->fixQuote($wordArray[2]), $this->fixQuote($wordArray[5]), 'NOT NULL');
+						}
+						break;
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php:171]: I think it is missing closing `}` to the switch statement.

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 8]
```diff
@@ -85,72 +85,86 @@ protected function buildCheckQuery()
...
+							$string = str_replace(';', '', $wordArray[8]);
+							$result = 'SELECT column_name, data_type, is_nullable FROM information_schema.columns WHERE table_name='
+								. $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+								. ' AND (CASE (position(' . $this->db->quote('::') . ' in column_default))'
+								. ' WHEN 0 THEN '
+							. '	column_default = ' . $this->db->quote($string)
+							. ' ELSE '
+							. ' substring(column_default, 1, (position('. $this->db->quote('::') . ' in column_default) -1))  = ' . $this->db->quote($string)
+							. ' END)';
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Add a blank line above.

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 9]
```diff
@@ -85,72 +85,86 @@ protected function buildCheckQuery()
...
-						. $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
-						. ' AND column_default ' . $isNullDef;
+					case 'DROP' :
+						if ($wordArray[7] === 'DEFAULT')
+						{
+							$isNullable = $this->fixQuote('NO');
+							$result = 'SELECT column_name, data_type, is_nullable , column_default FROM information_schema.columns'
+								. ' WHERE table_name=' . $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+								. ' AND column_default IS NULL';
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Add a blank line above.

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 10]
```diff
@@ -85,72 +85,86 @@ protected function buildCheckQuery()
...
-					$this->queryType = 'CHANGE_COLUMN_TYPE';
-					$this->checkQueryExpected = 1;
-					$this->msgElements = array($this->fixQuote($wordArray[2]), $this->fixQuote($wordArray[5]), $isNullDef);
+						if ($wordArray[7] . ' ' . $wordArray[8] === 'NOT NULL;')
+						{
+							$isNullable = $this->fixQuote('NO');
+							$result = 'SELECT column_name, data_type, is_nullable , column_default FROM information_schema.columns'
+								. ' WHERE table_name=' . $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+								. ' AND is_nullable = ' . $isNullable;
+							$this->queryType = 'CHANGE_COLUMN_TYPE';
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Add a blank line above.

File: libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php [hunk 11]
```diff
@@ -85,72 +85,90 @@ protected function buildCheckQuery()
...
+						}
+
+						if (strtoupper($wordArray[6] . ' ' . $wordArray[7]) === 'SET DEFAULT')
+						{
+							$string = str_replace(';', '', $wordArray[8]);
+							$result = 'SELECT column_name, data_type, is_nullable FROM information_schema.columns WHERE table_name='
+								. $this->fixQuote($wordArray[2]) . ' AND column_name=' . $this->fixQuote($wordArray[5])
+								. ' AND (CASE (position(' . $this->db->quote('::') . ' in column_default))'
+								. ' WHEN 0 THEN '
+							. '	column_default = ' . $this->db->quote($string)
```
Comment (Quy) [inline libraries/src/Schema/ChangeItem/PostgresqlChangeItem.php]: Indent lines 136-139.

**Comment by alikon** [comment]
sorry & thanks @Quy for your extra-work still fighting with the local setup of phpcs :face_with_head_bandage:

**Comment by alikon** [comment]
please  test #19707

**Pull request closed without merging**